### PR TITLE
Support custom dialers in dns registrar

### PR DESCRIPTION
**Pull request by mingyech**
Custom dialers are needed for Psiphon to route registration traffic outside of proxy. 

- [x] support custom TCP dialer for DoT and DoH registration
- [x] support custom UDP dialer for UDP registration
  - [x] refactor UDP requester to use `net.Conn` instead of `net.PacketConn`
- [x] dial registrar connection on first call to registration instead of on construction
- [x] use dialer from `ConjureSession` upon calls to register

### CODE REVIEW

**Review by jmwample — COMMENTED** [review]
I think there is an issue of actually providing the custom dialer function to the `Requester` in the constructor for the `DNSRegistrar` right now which is what a downstream library would use to build a `DNSRegistrar` to pass to the Conjure dialer. The downstream library should be able to pass a custom dialer that the DNS registrar will then give to the `Requester` that it builds. I can think of at least two ways to go about doing this, if you have a better option I am open to it.

1. We could add some kind of `NewDNSRegistrarWithDialContext` to `pkg/registration/dns-regstrar.go`
2. We could have the `(r *DNSRegistrar) register*directional(...)` functions call a dial function for the `Requester` using the dialer in the ConjureSession that is provided. This is the way that the API Registrar is able to use a custom dialer for creating its Connections. And while the dialers (all over actually) are named TcpDialer - that name is actually probably misleading (mistake on my part) because they are typically the `func (context.Context, network string, address string) (net.Conn, error)` which is totally fine to use `"udp"` as the network.

I would lean towards the second solution just to stay consistent with the APIRegistrar and the DecoyRegistrar which use the dialer from the session - which the downstream library can provide in the [Dialer](https://github.com/refraction-networking/gotapdance/blob/e3cad8ff1431cabae8c3fa0b54112ceab9204371/tapdance/dialer.go#L20). If we want to rename `tcpDialer` to something like `dialer` or `dialFunc` (same for `TcpDialer`) to avoid confusion going forward I am all for it.

So to sum up I would say we need to figure out:
- [ ] Only open connections once `Register(...)` has been called (not on construction)
- [ ] How does a downstream library provide a custom dialer to `NewDNSRegistrar` that will make it into the `Requester`?

File: pkg/dns-registrar/requester/requester.go [hunk 1]
```diff
@@ -36,13 +42,24 @@ func NewDoTRequester(dotaddr string, domain string, pubkey []byte, utlsDistribut
...
+			}
+			return tls.Client(conn, &tls.Config{}), nil
+		}
 	} else {
 		dialTLSContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
-			return utlsDialContext(ctx, network, addr, nil, utlsClientHelloID)
+			return utlsDialContext(ctx, network, addr, nil, utlsClientHelloID, dialContext)
 		}
 	}
 	dotconn, err := NewTLSPacketConn(dotaddr, dialTLSContext)
```
[the remark this answers is not in the document]
Comment (mingyech) [reply pkg/dns-registrar/requester/requester.go:65]: Good point, I agree that it is better to open the connection after we call the register function. It was set to call from the constructor to make the registrar reuse the connection when retrying the registration, but I think we can probably do that another way by keeping the transport established during the first registration.

File: pkg/dns-registrar/requester/requester.go [hunk 2]
```diff
@@ -100,18 +124,29 @@ func NewDoHRequester(dohurl string, domain string, pubkey []byte, utlsDistributi
...
 
-// New Requester using plain UDP as transport
+// New Requester using UDP as transport with default dialer
 func NewUDPRequester(remoteAddr net.Addr, domain string, pubkey []byte) (*Requester, error) {
+	listener := net.ListenConfig{}
+	return NewUDPRequesterWithListenPacket(remoteAddr, domain, pubkey, listener.ListenPacket)
+}
+
+// New Requester using UDP as transport
+func NewUDPRequesterWithListenPacket(remoteAddr net.Addr, domain string, pubkey []byte, listenPacket func(ctx context.Context, network, address string) (net.PacketConn, error)) (*Requester, error) {
```
[the remark this answers is not in the document]
Comment (mingyech) [reply pkg/dns-registrar/requester/requester.go:134]: The issue is that `UDPConn.WriteTo` that implements `net.PacketConn` does not work when it is "dialed" and seems to only work if it is "listened". I did try using the dialer with `"udp"` as network type and asserting it to a `net.UDPConn`, and there seems to be some internal state in `UDPConn` that keeps track of whether it was called via `dial` and returns an error on every call to `WriteTo`. `ListenPacket` was chosen because it has the closest function signature that I could find.

[the remark this answers is not in the document]
Comment (jmwample) [reply pkg/dns-registrar/requester/requester.go:134]: okay I was playing with this in an example program and the error that I was getting is something like `use of writeto with pre-connected connection` when trying to use a dialed `UDPConn` as a `PacketConn`. I haven't found a good way around this unfortunately. If we have to have a `listenPacket` function we should ensure a way for the downstream library to provide that function to the `DNSRegistrar` (either through the constructor or through the dialer/session/etc.). We might also want to add a `type ListenerContextFunc func(...)` so we are not passing around the function signature. 

It is a bummer that golang networking has to be like this.

[the remark this answers is not in the document]
Comment (jmwample) [reply pkg/dns-registrar/requester/requester.go:134]: One alternative would be to modify the `DNSRegistrar` to use a `UDPConn` instead of a `PacketConn`. That would allow us to use the same dial function signature, but we would need to modify the `recvLoop`  and `sendLoop` functions to use `Read` and `Write`.

Alternatively we could add a second DNSRegistrar type that functions basically the same and has a different constructor so we could build using either `PacketConn` or `UDPConn`. 

Just providing some options here. lmk what you think sounds easier / cleaner.

**Comment by mingyech** [comment]
Ah yeah you are right, I missed actually creating a way for the registrar to get the dialer. Now the registrar reads from the dialer from the registration data, similar to the API registrar. 

The DNS `Requester` now creates the underlying UDP/DoT/DoH connection during the first call to `RequestAndRecv`, which means that we are now only dialing during the first registration in the DNS registrar.

The constructors for the DNS and API registrars had a lot of arguments, which is getting a bit too long (my fault). I've made them now take a config object, which should make it easier and lets us add additional config variables in the future without breaking downstream libraries.

File: pkg/dns-registrar/requester/requester.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mingyech) [reply pkg/dns-registrar/requester/requester.go]: I've now made the DNS registrar take a `Conn` instead of `PacketConn`, which lets us use our dialer with UDP. While doing it, I added some stub functions in `QueuePacketConn` to make it implement the `Conn` interface that the `DNSPacketConn` now expects. 

I think the reason I chose `PacketConn` initially is that the connectionless nature of `PacketConn` is similar to DNS, but for our use case, we just need to connect to one remote so a `Conn` is better. Therefore we can actually remove the `QueuePacketConn` dependency altogether, but I think we can do that in another PR as there shouldn't be any API changes.

**Review by jmwample — APPROVED** [review]
This looks great! I am ready to get this merged. We should note in the tag / release that the `Dialer` interface has changed and the `TCPDialer` option is now `Dialer`, but nothing has functionally changed about it (noting for myself).

We might also add an example or two to `pkg/dns-registrar/README.md` to demonstrate how the `(New)DNSRegistrar` can/should be used and demo some of the key options (and the added defaults in `assets`).  Doesn't have to be part of this PR.
